### PR TITLE
release-23.2: rangefeed: fix premature checkpoint due to intent resolution race

### LINKED ISSUE #104309: rangefeed: Pushtxn in rangefeed returned abort, but txn may have been committed
I don't know if it's appropriate to send it here, but I'm really troubled.

The RangeFeed relies on the return result of PushTxns (task. go: pushOldTxns) ABORTED to remove the tracked txn in UnresolvedIntentQueue, which depends on the correctness of PushTxn.

https://github.com/cockroachdb/cockroach/blob/9c2f65094e2aa121d8e9b15c90e4d657848f1667/pkg/kv/kvserver/rangefeed/task.go#L324-L336

 But in cmd_push_txn.go The “case txnID” case of [PushTxn ->SynthesizeTxnFromMeta ->CanCreateTxnRecord(replica_tscache.go)]  may return a transaction that has already been COMMITTED but has status=Abort, which may cause rts in the RangeFeed to advance incorrectly. Do I understand correctly?
https://github.com/cockroachdb/cockroach/blob/9c2f65094e2aa121d8e9b15c90e4d657848f1667/pkg/kv/kvserver/replica_tscache.go#L529-L543

Jira issue: CRDB-28452

Epic CRDB-27235

### PR DESCRIPTION
**Pull request by erikgrinaker**
Only the 2 last commits should be reviewed here, the earlier commits in this PR are from the following prerequisite backports:

* #118407
* #118409
* #118411

This backport includes new RPC protocol additions that are necessary to fix the bug. The new behavior (applying a `Barrier` command to wait for a Raft pipeline flush) is controlled by a cluster setting `kv.rangefeed.push_txns.barrier.enabled`, which defaults to `true`. According to the backport policy, such changes must be opt-in and disabled by default. However, making a bug fix opt-in appears questionable, especially one as serious as this one. I consider the protocol changes to be low-risk high-reward, and recommend we backport this behavior enabled by default.

@rharding6373 Assigning you as secondary TL reviewer, since you're the upcoming CDC TL. Let me know if you'd like me to route this elsewhere.

---

Backport 2/2 commits from https://github.com/cockroachdb/cockroach/pull/117612. Also pull in parts of #118469 and #118265.

Release justification: fixes a bug which could cause changefeeds to omit events in some scenarios.

/cc @cockroachdb/release

---

It was possible for rangefeeds to emit a premature checkpoint, before all writes below its timestamp had been emitted. This in turn would cause changefeeds to not emit these write events at all.

This could happen because `PushTxn` may return a false `ABORTED` status for a transaction that has in fact been committed, if the transaction record has been GCed (after resolving all intents). The timestamp cache does not retain sufficient information to disambiguate a committed transaction from an aborted one in this case, so it pessimistically
assumes an abort (see `Replica.CanCreateTxnRecord` and `batcheval.SynthesizeTxnFromMeta`).

However, the rangefeed txn pusher trusted this `ABORTED` status, ignoring the pending txn intents and allowing the resolved timestamp to advance past them before emitting the committed intents. This can lead to the following scenario:

- A rangefeed is running on a lagging follower.
- A txn writes an intent, which is replicated to the follower.
- The closed timestamp advances past the intent.
- The txn commits and resolves the intent at the original write timestamp, then GCs its txn record. This is not yet applied on the follower.
- The rangefeed pushes the txn to advance its resolved timestamp.
- The txn is GCed, so the push returns ABORTED (it can't know whether the txn was committed or aborted after its record is GCed).
- The rangefeed disregards the "aborted" txn and advances the resolved timestamp, emitting a checkpoint.
- The follower applies the resolved intent and emits an event below the checkpoint, violating the checkpoint guarantee.
- The changefeed sees an event below its frontier and drops it, never emitting these events at all.

This patch fixes the bug by submitting a barrier command to the leaseholder which waits for all past and ongoing writes (including intent resolution) to complete and apply, and then waits for the local replica to apply the barrier as well. This ensures any committed intent resolution will be applied and emitted before the transaction is removed from resolved timestamp tracking.

Resolves #104309.
Epic: none
Release note (bug fix): fixed a bug where a changefeed could omit events in rare cases, logging the error "cdc ux violation: detected timestamp ... that is less or equal to the local frontier". This can happen if a rangefeed runs on a follower replica that lags significantly behind the leaseholder, a transaction commits and removes its transaction record before its intent resolution is applied on the follower, the follower's closed timestamp has advanced past the transaction commit timestamp, and the rangefeed attempts to push the transaction to a new timestamp (at least 10 seconds after the transaction began). This may cause the rangefeed to prematurely emit a checkpoint before emitting writes at lower timestamps, which in turn may cause the changefeed to drop these events entirely, never emitting them.

### CODE REVIEW

**Comment by erikgrinaker** [comment]
[Approved by CTO](https://cockroachlabs.slack.com/archives/C060S8QKABT/p1706536509509529?thread_ts=1706535112.576979&cid=C060S8QKABT).

File: pkg/kv/kvserver/rangefeed/resolved_timestamp.go [hunk 1]
```diff
@@ -265,10 +275,20 @@ func (rts *resolvedTimestamp) assertNoChange() {
...
-func (rts *resolvedTimestamp) assertOpAboveRTS(op enginepb.MVCCLogicalOp, opTS hlc.Timestamp) {
+func (rts *resolvedTimestamp) assertOpAboveRTS(
+	ctx context.Context, op enginepb.MVCCLogicalOp, opTS hlc.Timestamp, fatal bool,
+) {
 	if opTS.LessEq(rts.resolvedTS) {
-		panic(fmt.Sprintf("resolved timestamp %s equal to or above timestamp of operation %v",
-			rts.resolvedTS, op))
+		// NB: MVCCLogicalOp.String() is only implemented for pointer receiver.
+		err := errors.AssertionFailedf(
+			"resolved timestamp %s equal to or above timestamp of operation %v", rts.resolvedTS, &op)
```
Comment (nvanbenschoten) [inline pkg/kv/kvserver/rangefeed/resolved_timestamp.go:284]: We're now passing `op` by reference. Does that cause the argument to escape to the heap and result in a new heap allocation even when the assertion is not firing?

**Review by rharding6373 — APPROVED** [review]
:lgtm: Thanks for all the work finding, fixing, and backporting the fixes for this issue!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118413#-:-NpM7Jll37z9fQmTf3XB:b-7xl944)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)

File: pkg/kv/kvserver/rangefeed/resolved_timestamp.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/rangefeed/resolved_timestamp.go:284]: Great catch, especially considering I slipped this in here from #118265 -- thanks. You're probably right, I'll check this tomorrow and either shadow it in the branch or revert to passing by value.

**Review by nvb — APPROVED** [review]
LGTM once the discussion from https://github.com/cockroachdb/cockroach/pull/118265 is resolved.

**Comment by yuzefovich** [comment]
Just a heads up that I think this currently is not on 23.2.1-rc branch.

**Comment by erikgrinaker** [comment]
Thanks for the headsup, will bump it over.

**Pull request merged**